tools/msu_finder.rb

```diff
@@ -0,0 +1,760 @@
+    #
+    # @param msb [String] MSB number in this format: msxx-xxx
+    # @return [Rex::Proto::Http::Response]
+    def download_advisory(msb)
+      send_http_request(SiteInfo::TECHNET, {
+        'uri' => "/en-us/library/security/#{msb}.aspx"
+      })
+    end
+
+
+    # Returns the most appropriate pattern that could be used to parse and extract links from an advisory.
+    #
+    # @param n [Nokogiri::HTML::Document] The advisory page parsed by Nokogiri
+    # @return [Hash]
+    def get_appropriate_pattern(n)
+      # These pattern checks need to be in this order.
+      patterns = [
+        # This works from MS14-001 until the most recent
+        {
+          check:   '//div[@id="mainBody"]//div//h2//div//span[contains(text(), "Affected Software")]',
+          pattern: '//div[@id="mainBody"]//div//div[@class="sectionblock"]//table//a' 
+        },
+        # This works from ms03-040 until MS07-029
+        {
+          check:   '//div[@id="mainBody"]//ul//li//a[contains(text(), "Download the update")]',
+          pattern: '//div[@id="mainBody"]//ul//li//a[contains(text(), "Download the update")]'
+        },
+        # This works from sometime until ms03-039
+        {
+          check:   '//div[@id="mainBody"]//div//div[@class="sectionblock"]//p//strong[contains(text(), "Download locations")]',
+          pattern: '//div[@id="mainBody"]//div//div[@class="sectionblock"]//ul//li//a'
+        },
+        # This works from MS07-030 until MS13-106 (the last update in 2013)
+        # The check is pretty short so if it kicks in too early, it tends to create false positives.
+        # So it goes last.
+        {
+          check:   '//div[@id="mainBody"]//p//strong[contains(text(), "Affected Software")]',
+          pattern: '//div[@id="mainBody"]//table//a' 
+        },
+      ]
+
+      patterns.each do |pattern|
+        if n.at_xpath(pattern[:check])
+          return pattern[:pattern]
+        end
+      end
+
+      nil
+    end
+
+
+    # Returns the details page for an advisory.
+    #
+    # @param res [Rex::Proto::Http::Response]
+    # @return [Array<URI::HTTP>] An array of URI objects.
+    def get_details_aspx(res)
+      links = []
+
+      page = res.body
+      n = ::Nokogiri::HTML(page)
+
+      appropriate_pattern = get_appropriate_pattern(n)
+
+      n.search(appropriate_pattern).each do |anchor|
+        found_link = anchor.attributes['href'].value
+        if /https:\/\/www\.microsoft\.com\/downloads\/details\.aspx\?familyid=/i === found_link
+          begin
+            links << URI(found_link)
+          rescue ::URI::InvalidURIError
+            print_error "Unable to parse URI: #{found_link}"
+          end
+        end
+      end
+
+      links
+    end
+
+
+    # Returns the redirected page.
+    #
+    # @param rhost [Hash] From MicrosoftPatch::SiteInfo
+    # @param res [Rex::Proto::Http::Response]
+    # @return [Rex::Proto::Http::Response]
+    def follow_redirect(rhost, res)
```

Comment:
I'd say this method could be reused by others, since looks just related to http, not specific from MSU downloads. Maybe worths to have a mixin / class for HTTP things? `Base` also includes a `send_http_request` method.